ARROW-9630: [Go] Implement public JSON reader/writer

zeroshade opened the pull request:
This JSON reader/writer is intended to support the standard Go encoding/json package for Go and writes out arrays and records as data, not as the integration format. This implements `MarshalJSON` for the array types, and `UnmarshalJSON` for the array builders (in order to unmarshal json data to add to an existing builder) along with providing a few helper functions for single array from json or single recordbatch from json logic.

emkornfield commented on go/arrow/array/binary.go:
```diff
@@ -117,6 +118,21 @@ func (a *Binary) setData(data *Data) {
+	for i := 0; i < a.Len(); i++ {
+		vals[i] = a.getOneForMarshal(i)
+	}
+	return json.Marshal(vals)
```
How do non-printable characters get marshalled in json?  Escaped bytes?

emkornfield replied:
Okay, based on below is looks like base64 encoded?  This is not something I could find as a reference in the docs?

zeroshade replied:
From the Go docs on JSON marshalling: https://pkg.go.dev/encoding/json#Marshal

> Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON value.

The standard marshalling of byte slices to JSON in go will marshal it to a base64-encoded string.

emkornfield commented on go/arrow/array/decimal128.go:
```diff
@@ -229,6 +253,67 @@ func (b *Decimal128Builder) newData() (data *Data) {
+	case float64:
+		out = big.NewFloat(v)
+	case string:
+		out, _, err = big.ParseFloat(v, 10, 0, big.ToNearestAway)
```
any strong rationale for ToNearestAway?  maybe document?

zeroshade replied:
No strong rationale, just happened to be what got me the closest to the value I had input when I used it and I couldn't think of a good way to easily make this configurable in the current interface. I'll add a comment to document it.

zeroshade replied:
comments added

emkornfield commented on go/arrow/array/decimal128.go:
```diff
@@ -229,6 +253,67 @@ func (b *Decimal128Builder) newData() (data *Data) {
+	case float64:
+		out = big.NewFloat(v)
+	case string:
+		out, _, err = big.ParseFloat(v, 10, 0, big.ToNearestAway)
```
any strong rationale for ToNearestAway?  maybe document?

zeroshade replied:
duplicate comment of https://github.com/apache/arrow/pull/11359#discussion_r741359958

emkornfield replied:
Also, is 0 the right value for precision here shouldn't it be 128?

Also if I read the docs correctly it seems that this will silently truncate decimal values that require larger precision, do you have thoughts on if it would be better to check for these types of values and return an error instead?

emkornfield commented on go/arrow/array/decimal128.go:
```diff
@@ -229,6 +253,67 @@ func (b *Decimal128Builder) newData() (data *Data) {
+		if err != nil {
+			return err
+		}
+	case json.Number:
```
I might be misunderstanding but it seems like json.Number is an alias for string (https://pkg.go.dev/encoding/json#Number).  Should this go through the same logic above for string?

zeroshade replied:
It's not an alias but an actual type, which would only get used if the json encoder sets an option to use it. The `json.Number` type provides helper functions for parsing from the string, so the logic wouldn't be the same as using string, but instead using the helper functions directly and checking the error returns from them

emkornfield replied:
But it like accessing the [string literal](https://pkg.go.dev/encoding/json#Number.String) is supposed.  It seems like Float64 decoding loses precision needlessly

zeroshade replied:
good point, i've updated the `json.Number` case to work the same as the `string` case and fixed the precision to be 128

emkornfield commented on go/arrow/array/fixed_size_list.go:
```diff
@@ -110,6 +112,44 @@ func (a *FixedSizeList) Release() {
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+
+	buf.WriteByte('[')
```
Is this using a string to encode FixedSizeLists?  That might be counter-intuitive to people?  I guess this is needed to round-trip with full-fidelity?

zeroshade replied:
No, it's just encoding a fixed size list identically to how it encodes regular lists, as a nested list. From the perspective of JSON there's no difference between a `List` where all of the lists happen to be the same length, and a `FixedSizeList`. A fixed size list with `n=3` of Int would encode in json as `[ [1,2,3], [4,5,6], [7,8,9] ]`

zeroshade replied:
I can understand the confusion now that I think about it. The return from `MarshalJSON` is the byte slice of the raw JSON that is produced by the conversion. I hope that clears things up a bit.

emkornfield commented on go/arrow/array/float16_builder.go:
```diff
@@ -163,3 +168,59 @@ func (b *Float16Builder) newData() (data *Data) {
+		if err != nil {
+			return err
+		}
+		b.Append(float16.New(float32(f)))
```
Does this silently truncate or does it throw an error?

zeroshade replied:
Currently it will silently truncate

zeroshade replied:
comment added to document the truncation

emkornfield replied:
OK, we should probably also make this clear in public docs.

zeroshade replied:
It's in the docstring for the `UnmarshalJSON` function which means it'll appear in the public docs for the `UnmarshalJSON` on `Float16Builder`

emkornfield commented on go/arrow/array/float16_builder.go:
```diff
@@ -163,3 +168,59 @@ func (b *Float16Builder) newData() (data *Data) {
+	}
+
+	if delim, ok := t.(json.Delim); !ok || delim != '[' {
+		return fmt.Errorf("binary builder must unpack from json array, found %s", delim)
```
check for copy and paste but in error messages?

emkornfield commented on go/arrow/array/interval.go:
```diff
@@ -86,6 +88,29 @@ func (a *MonthInterval) setData(data *Data) {
+	vals := make([]interface{}, a.Len())
+	for i := 0; i < a.Len(); i++ {
+		if a.IsValid(i) {
+			vals[i] = a.values[i]
```
In this case, it seems like it might be better to have a formatted string (like "1m") or "1months"

emkornfield replied:
sorry, not seeing this appear (maybe forgot to push?)

zeroshade replied:
It was handled by the change in the `datatype_fixedwidth.go` file by adding `MarshalJSON` and `UnmarshalJSON` functions to the `MonthInterval` type here: https://github.com/apache/arrow/pull/11359/files#diff-ada6f476fe71c04e54cb986f027d91bad2517d3a5a1f278115a80ea3bcc18482R333 When it goes to marshal the array of `MonthInterval` values, it'll use the `MarshalJSON` and `UnmarshalJSON` functions defined on the type, rather than defaulting to just treating it like an integer.

emkornfield commented on go/arrow/array/interval.go:
```diff
@@ -279,6 +341,28 @@ func (a *DayTimeInterval) setData(data *Data) {
+
+func (a *DayTimeInterval) MarshalJSON() ([]byte, error) {
+	if a.NullN() == 0 {
+		return json.Marshal(a.values)
```
same comment, as months, consider making a formatted string that can be reparsed?  I guess whether we use a formatted string or the raw values comes down to who we expect to consume the JSON.  I think for machines the numbers might be more natural, but for human consumption formatted string makes it clearer what the types are.

zeroshade replied:
Currently DayTime interval would get marshaled as `{ "days": #, "milliseconds": # }`, the same would be true for month day nano interval type being marshalled as `{ "months": #, "days": #, "nanoseconds": # }`. The only one that currently wouldn't get marshalled into a formatted object or string is the month interval. The intention I had was that the actual types should be defined externally before attempting to parse the JSON as I didn't build any arrow type inference into the JSON reader. It requires being passed a schema or data type.

emkornfield replied:
SGTM, might be worth documenting what the value looks like here.

zeroshade replied:
Just to add on since my last comment, I've since modified the Month Interval to marshal as a string like you suggested as "1months" etc. but sure, i'll add a specific comment on these marshal functions about their format so it's not only on the top level "to json" functions.

zeroshade replied:
comments added.

emkornfield commented on go/arrow/array/map.go:
```diff
@@ -266,6 +270,28 @@ func (b *MapBuilder) ValueBuilder() *StructBuilder {
+	}
+
+	if delim, ok := t.(json.Delim); !ok || delim != '[' {
+		return fmt.Errorf("binary builder must unpack from json array, found %s", delim)
```
copy-paste error message?

emkornfield commented on go/arrow/array/numeric.gen.go:
```diff
@@ -80,6 +82,27 @@ func (a *Int64) setData(data *Data) {
+		return nil
+	}
+
+	return float64(a.values[i]) // prevent uint8 from being seen as binary data
```
won't this truncate int64 values that require > 52 bits of storage?

zeroshade replied:
Fair, I'll update the template to only add that to the uint8 type

zeroshade replied:
fixed

emkornfield commented on go/arrow/array/numeric.gen.go:
```diff
@@ -750,6 +962,22 @@ func (a *Timestamp) setData(data *Data) {
+	if a.IsNull(i) {
+		return nil
+	}
+	return a.values[i].ToTime(a.DataType().(*arrow.TimestampType).Unit).Format("2006-01-02 15:04:05.999999999")
```
should formatting be conditional on precision?

zeroshade replied:
using `9` there will leave out trailing zeros from the resulting string so this will handle whatever level of precision is provided rather than explicitly padding to a specific precision level.

emkornfield commented on go/arrow/array/numeric.gen.go:
```diff
@@ -817,6 +1045,22 @@ func (a *Time32) setData(data *Data) {
+	if a.IsNull(i) {
+		return nil
+	}
+	return a.values[i].ToTime(a.DataType().(*arrow.Time32Type).Unit).Format("15:04:05.999999999")
```
since it is a different type, it is probably easy to format this to only the precision required?

zeroshade replied:
using the `9`'s there will drop any trailing zeros so that it will only format it to the precision required to represent the value.

emkornfield commented on go/arrow/array/numeric.gen.go:
```diff
@@ -1085,6 +1377,22 @@ func (a *Duration) setData(data *Data) {
+	if a.IsNull(i) {
+		return nil
+	}
+	return fmt.Sprint(time.Duration(a.values[i]) * a.DataType().(*arrow.DurationType).Unit.Multiplier())
```
this logic is a little opaque to me, would you mind adding a comment on intended output.

emkornfield replied:
Are there overflow issues for courser precisions?

zeroshade replied:
Yes, there can end up being overflow issues for large enough values depending on the precision. I'll see if i can tweak this to try to avoid the overflows.

zeroshade replied:
fixed

emkornfield commented on go/arrow/array/struct.go:
```diff
@@ -105,6 +107,36 @@ func (a *Struct) setData(data *Data) {
+	tmp := make(map[string]interface{})
+	fieldList := a.data.dtype.(*arrow.StructType).Fields()
+	for j, d := range a.fields {
+		tmp[fieldList[j].Name] = d.getOneForMarshal(i)
```
I don't think the arrow spec says struct names need to be unique, it might be worth checking for that case here and throwing an error OR using a list of tuple representation in that case.

zeroshade replied:
currently the Go implementation does not support duplicate field names in structs, only in schemas. So at least for now i'm guaranteed that there will be no duplicate field names until i make a change to support them in the struct type in general.

emkornfield replied:
hmm, OK, I might be wrong about the struct then, I might have been thinking about schema.

emkornfield commented on go/arrow/array/struct.go:
```diff
@@ -105,6 +107,36 @@ func (a *Struct) setData(data *Data) {
+	enc := json.NewEncoder(&buf)
+
+	buf.WriteByte('[')
+	for i := 0; i < a.Len(); i++ {
```
I might be getting a little lost here but IIUC the other json functions return a list representing the column value.  Here it seems we transpose each column to a row first?  This seems to lack symmetry?  Maybe in this case the MarshalJSON function shouldn't use getOneForMarshal?

zeroshade replied:
In a struct column, each "value" in the list, is an object consisting of keys from the struct field names with values for the corresponding index. 

For example, a `struct<a: int, b: float>` column would be marshalled to `[ {"a": <a[0]>, "b": <b[0]>}, {"a": <a[1]>, "b": <b[1]> }, .... ]`. Does that make more sense?

emkornfield replied:
Yeah, so I think I've come around to how this is implemented.  I think what I was struggling with is an IntArray, gets serialized as `[1,2,3]` but one could maybe think it should get serialized as something like `[[1],[2],[3]` or `[{'entry':1}, {'entry':2}, {'entry': 3]` similar to what struct does).  I guess it is good if someone just wants to serialize/deserialize just an array and not a record batch that it happens as a flat JSON list and the row by row approach is taken for struct.

emkornfield commented on go/arrow/array/util.go:
```diff
@@ -16,9 +16,178 @@
+// The JSON provided must be formatted in one of two ways:
+//		Default: the top level of the json must be a list which matches the type specified exactly
+//		Example: `[1, 2, 3, 4, 5]` for any integer type or `[[...], null, [], .....]` for a List type
+//					Struct arrays are represented a list of objects: `[{"foo": 1, "bar": "moo"}, {"foo": 5, "bar": "baz"}]`
```
Maybe also document time/interal/duration types?

zeroshade replied:
added documentation for those types

emkornfield commented on go/arrow/array/util_test.go:
```diff
@@ -0,0 +1,406 @@
+		valueadd int
+	}{
+		{arrow.FixedWidthTypes.Time32s, `[null, "10:10:10"]`, 123},
+		{arrow.FixedWidthTypes.Time32ms, `[null, "10:10:10.123"]`, 456},
```
oh is the formatter smart to not extend 0 values on the decimal?

zeroshade replied:
Right, that's the difference between using 0's or 9's in the format string. Using 0's will extend it to the exact precision stated, using 9's will truncate trailing zeros.

emkornfield commented on go/arrow/datatype_fixedwidth.go:
```diff
@@ -54,6 +56,140 @@ type (
+	case Second:
+		return Timestamp(out.Unix()), nil
+	case Millisecond:
+		return Timestamp(out.Unix()*1e3 + int64(out.Nanosecond())/1e6), nil
```
I know I asked this before but Go panics on overflow?

emkornfield commented on go/arrow/datatype_fixedwidth.go:
```diff
@@ -54,6 +56,140 @@ type (
+	case Second:
+		return Timestamp(out.Unix()), nil
+	case Millisecond:
+		return Timestamp(out.Unix()*1e3 + int64(out.Nanosecond())/1e6), nil
```
I think at least in C++ errors are raised when if precision is truncated (i.e. we'd expect everything past the millisecond precision to be 0)

zeroshade replied:
Looking at https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/value_parsing.h#L530 it looks like the C++ doesn't even accept the excess 0s. It errors if more than the right number of decimal places are provided. So i should be able to easily handle the error cases by using the unit and checking the length of the string. I'll add the errors and update the tests.

emkornfield reviewed (changes requested):
I think this will be really useful.  A few of the comments I left (JSON marshalling of struct type) and behavior on truncation are a little philosophical, I think for data-processing systems being conservative and by default erroring on truncation can be less surprising for users later  on when they try to figure out why precision is missing, but it is in some ways less userfriendly for something as free-form as JSON.  It doesn't seem like there is a great way to control the behavior through the provided go JSON marshaling APIs so we need to pick one, either way we should clearly document the choice and see if this is maybe controllable at the Reader level

zeroshade commented:
@emkornfield I've addressed the comments you left either via comments or via changes. I tend to lend more towards being less unfriendly wherever possible rather than letting it fail personally which is why I took the tact I did in a lot of the places, but I also see the value in matching behavior of the C++ library in a lot of this since there isn't really a published "spec" for the a JSON form of Arrow it's hard to standardize. Let me know what you think of the changes I've made and comments/documentation I've added.

emkornfield commented:
> I've addressed the comments you left either via comments or via changes. I tend to lend more towards being less unfriendly wherever possible rather than letting it fail personally which is why I took the tact I did in a lot of the places, but I also see the value in matching behavior of the C++

It's not so much about matching C++, as it is about users getting unexpected results.  I don't know if there is a good way to disclose some of caveats in a way a user will actually look at before blindly going ahead and using the library.  For now clear documentation for both reading and writing that data is attempted to be coerced as best as possible to the schema provided and might lose some data from the original JSON is fine.  We can re-assess if anyone opens up bugs because of this.

zeroshade commented:
@emkornfield I've updated based on the feedback. I'll wait until I get an approval from you to merge unless you have more comments or something i missed. Thanks again for the review!

emkornfield commented:
Thanks, I'll review the updates.  It occurred to me that for Interval months it should maybe serialize as `{"months": X}` which would allow import as months and also potentially as `MonthDayNano`.  Sorry for the churn here and feel free for ignore this.

emkornfield commented on go/arrow/array/interval.go:
```diff
@@ -95,6 +95,9 @@ func (a *MonthInterval) getOneForMarshal(i int) interface{} {
 	return nil
 }
 
+// MarshalJSON will create a json array out of a MonthInterval array,
```
not sure if you saw my last comment what do you think about thins`{months=num}`

zeroshade replied:
I saw the comment, I wasn't feeling too strongly one way or the other but sure, I'll bump it to that format since it's easy enough to do. After everything passes I can commit this, if you're willing to set your review to approve the changes :)

zeroshade replied:
latest commit changes the format to `{"months": #}` and I updated the comments to reflect that in the documentation.

emkornfield commented:
LGTM modulo the months interval comment which I don't feel too strongly about.

emkornfield reviewed (approved):
I think you need to rebase.

zeroshade commented:
@emkornfield  Yup, just did the rebase. Once everything passes i'll merge. Thanks for the review and comments!
